botocore/auth.py

```diff
@@ -362,12 +363,52 @@ def _set_necessary_date_headers(self, request):
 
 
 class S3SigV4Auth(SigV4Auth):
+    def __init__(self, credentials, service_name, region_name):
+        super(S3SigV4Auth, self).__init__(
+            credentials, service_name, region_name)
+        self._default_region_name = region_name
+
+    def add_auth(self, request):
+        # If we ever decide to share auth sessions, this could potentially be
+        # a source of concurrency bugs.
+        signing_context = request.context.get('signing', {})
+        self._region_name = signing_context.get(
+            'region', self._default_region_name)
+        super(S3SigV4Auth, self).add_auth(request)
 
     def _modify_request_before_signing(self, request):
         super(S3SigV4Auth, self)._modify_request_before_signing(request)
         if 'X-Amz-Content-SHA256' in request.headers:
             del request.headers['X-Amz-Content-SHA256']
-        request.headers['X-Amz-Content-SHA256'] = self.payload(request)
+
+        if self._should_sha256_sign_payload(request):
+            request.headers['X-Amz-Content-SHA256'] = self.payload(request)
+        else:
+            request.headers['X-Amz-Content-SHA256'] = 'UNSIGNED-PAYLOAD'
+
+    def _should_sha256_sign_payload(self, request):
+        # S3 allows optional body signing, so to minimize the performance
+        # impact, we opt to not SHA256 sign the body on streaming uploads,
+        # provided that we're on https.
+        client_config = request.context.get('client_config')
+        s3_config = getattr(client_config, 's3', None)
+
+        # The config could be None if it isn't set, or if the customer sets it
+        # to None.
+        if s3_config is None:
+            s3_config = {}
+
+        use_sha256 = s3_config.get('sha256_sign_s3v4_payload')
```

Comment:
Hmm. So I think we can take the ``s3`` part out at least since it is in the s3 config already. Honestly, I do not have a strong opinion on what the name should be. People will have to read the docs anyway to understand what it does so as long as we are clear there I am fine.

As to ``sign_payloads``, if there is a new auth, that comes along would that mean we have to not sign the payload there as well if the config option is turned off? It may be good to keep the sigv4 part in the name for backwards compatibility reasons.